src/e2e/app/common/src/e2e_test.rs

```diff
@@ -0,0 +1,149 @@
+
+        fixture_res.int_err()
+    };
+
+    let (server_output, test_res) = tokio::join!(server_run_fut, test_fut);
+
+    if let Err(e) = test_res {
+        let mut panic_message = format!("Fixture execution error:\n{e:?}\n");
+
+        panic_message += "Server output:\n";
+        panic_message += "stdout:\n";
+        panic_message += server_output.stdout.as_str();
+        panic_message += "\n";
+        panic_message += "stderr:\n";
+        panic_message += server_output.stderr.as_str();
+        panic_message += "\n";
+
+        panic!("{panic_message}");
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub async fn api_flight_sql_e2e_test<ServerRunFut, Fixture, FixtureFut>(
+    e2e_data_file_path: PathBuf,
+    workspace_path: PathBuf,
+    server_run_fut: ServerRunFut,
+    fixture: Fixture,
+) where
+    ServerRunFut: Future<Output = ServerOutput>,
+    Fixture: FnOnce(KamuFlightSQLClient) -> FixtureFut,
+    FixtureFut: Future<Output = ()> + Send + 'static,
+{
+    let test_fut = async move {
+        let base_url = get_server_api_base_url(e2e_data_file_path.clone()).await?;
+        let kamu_api_server_client = KamuApiServerClient::new(base_url, workspace_path);
+        let flight_sql_base_url = get_server_flight_sql_base_url(e2e_data_file_path).await?;
+
+        let kamu_flight_sql_server_client = kamu_api_server_client
+            .flight_sql_client(flight_sql_base_url)
+            .await;
+
+        kamu_api_server_client.e2e().ready().await?;
+
+        let fixture_res = {
+            // tokio::spawn() is used to catch panic, otherwise the test will hang
+            tokio::spawn(fixture(kamu_flight_sql_server_client.clone())).await
+        };
+
+        kamu_api_server_client.e2e().shutdown().await?;
+
+        fixture_res.int_err()
+    };
+
+    let (server_output, test_res) = tokio::join!(server_run_fut, test_fut);
+
+    if let Err(e) = test_res {
+        let mut panic_message = format!("Fixture execution error:\n{e:?}\n");
+
+        panic_message += "Server output:\n";
+        panic_message += "stdout:\n";
+        panic_message += server_output.stdout.as_str();
+        panic_message += "\n";
+        panic_message += "stderr:\n";
+        panic_message += server_output.stderr.as_str();
+        panic_message += "\n";
+
+        panic!("{panic_message}");
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+async fn get_server_api_base_url(e2e_data_file_path: PathBuf) -> Result<Url, InternalError> {
+    let retry_strategy = FixedInterval::from_millis(500).take(10);
+    let base_url = Retry::spawn(retry_strategy, || async {
+        let data = tokio::fs::read_to_string(e2e_data_file_path.clone())
+            .await
+            .int_err()?;
+
+        Url::parse(data.as_str().split_once('\n').unwrap().0).int_err()
```

Comment:
Thoughts: we need to either describe in the comments why we read only the first line (and then the second line).

Also, it would be nice, to improve readability, to extract the variable:

```rust
let (api_server_base_url, _flight_sql_base_url) = data.as_str().split_once('\n').unwrap();
```

Reply:
(in this and the following method)